### PR TITLE
Fix cache:rebuild running incorrectly on D7 sites.

### PR DESCRIPTION
**Pull request by markdorison**
## Description
Fix cache:rebuild running incorrectly on D7 sites.

## Motivation / Context
Error on D7 site.

## Testing Instructions / How This Has Been Tested
Tested manually.

cc/ @mmatsoo

### CODE REVIEW

**Review by mmatsoo — APPROVED** [review]
👍 I have not tested against a D7 site, but having gone through yesterday's Intuit challenges, the changes here look like they overcome the cache-clearing difficulties I discovered.

File: src/Robo/Plugin/Commands/DevelopmentModeBaseCommands.php [hunk 1]
```diff
@@ -124,10 +126,12 @@ public function databaseRefreshTugboat(): Result
...
-            $result = $this->taskExec("$this->vendorDirectory/bin/drush")
-                ->arg('cache:rebuild')
-                ->dir("$this->drupalRoot/sites/$siteName")
-                ->run();
+            if (!$this->drupalVersionIsD7($this->drupalRoot)) {
+                $result = $this->taskExec("$this->vendorDirectory/bin/drush")
+                    ->arg('cache:rebuild')
+                    ->dir("$this->drupalRoot/sites/$siteName")
+                    ->run();
+            }
```
Comment (adamzimmermann) [inline src/Robo/Plugin/Commands/DevelopmentModeBaseCommands.php:134]: Do we need an `else` for D7?

File: src/Robo/Plugin/Traits/DrupalVersionTrait.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ */
+trait DrupalVersionTrait
+{
+    /**
+     * Check if version of Drupal in question is Drupal 7.
+     *
+     * @param string $drupalRootPath
+     *   The path to the Drupal root.
+     */
+    protected function drupalVersionIsD7(string $drupalRootPath): bool
```
Comment (adamzimmermann) [inline src/Robo/Plugin/Traits/DrupalVersionTrait.php:16]: Could we do something like this?

- https://www.liquidweb.com/kb/version-drupal-running/

Seems slightly less brittle?

I don't want us to over think this though either if this works. 🤷🏼‍♂️

**Review by adamzimmermann — APPROVED** [review]
A few questions, but looks like it works as is. ✅

File: src/Robo/Plugin/Commands/DevelopmentModeBaseCommands.php (hunk 1 above)
Comment (markdorison) [reply src/Robo/Plugin/Commands/DevelopmentModeBaseCommands.php:134]: This was [added (relatively) recently](https://github.com/ChromaticHQ/usher/commit/4d02216282412f8133de6accee4152c87ffe9b5b) to solve a problem on D8/9/10. I didn't want to add a step for D7 if I didn't know that it was needed.

Comment (adamzimmermann) [reply src/Robo/Plugin/Commands/DevelopmentModeBaseCommands.php:134]: Ah, this is that cache clear. 👍🏼

File: src/Robo/Plugin/Traits/DrupalVersionTrait.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        // This line has not changed since 2013, so it seems safe to base our version check off of.
+        // https://git.drupalcode.org/project/drupal/-/blame/7.x/COPYRIGHT.txt
+        $copyrightString = 'All Drupal code is Copyright 2001 - 2013 by the original authors.';
+        $copyrightPath = "$drupalRootPath/COPYRIGHT.txt";
+        if (file_exists($copyrightPath) && $copyrightContents = file_get_contents($copyrightPath)) {
+            if (str_contains($copyrightContents, $copyrightString)) {
```
Comment (markdorison) [inline src/Robo/Plugin/Traits/DrupalVersionTrait.php]:
```suggestion
         // @see https://git.drupalcode.org/project/drupal/-/blob/7.x/includes/bootstrap.inc
        $bootstrapPath = "$drupalRootPath/includes/bootstrap.inc";
        if (file_exists($bootstrapPath)) {
            include_once($bootstrapPath);
            if (substr(VERSION, 0, 2) === '7.') {
```

@adamzimmermann Is this less brittle? Relying on the version number definitely *seems* more reliable on the surface, but the version constant changes value frequently vs the string that hasn't changed in 10 years. What do you think?

Comment (adamzimmermann) [reply src/Robo/Plugin/Traits/DrupalVersionTrait.php]: > but the version constant changes value frequently

Your check accounts for the change though, right?

Or is there another element of risk that I'm not considering?

Comment (markdorison) [reply src/Robo/Plugin/Traits/DrupalVersionTrait.php]: It _should_... I am probably not being trusting enough.

Comment (adamzimmermann) [reply src/Robo/Plugin/Traits/DrupalVersionTrait.php]: I always support being paranoid with code.

Comment (markdorison) [reply src/Robo/Plugin/Traits/DrupalVersionTrait.php]: Alright, I am going for it. 😆

File: src/Robo/Plugin/Traits/DrupalVersionTrait.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+     * @param string $drupalRootPath
+     *   The path to the Drupal root.
+     */
+    protected function drupalVersionIsD7(string $drupalRootPath): bool
+    {
+        // @see https://git.drupalcode.org/project/drupal/-/blob/7.x/includes/bootstrap.inc
+        $bootstrapPath = "$drupalRootPath/includes/bootstrap.inc";
+        if (file_exists($bootstrapPath)) {
+            include_once($bootstrapPath);
+            if (substr(VERSION, 0, 2) === '7.') {
```
Comment (markdorison) [inline src/Robo/Plugin/Traits/DrupalVersionTrait.php]: @adamzimmermann Ahh. PHPStan really doesn't like the use of this constant. I suppose I understand.

CleanShot 2023-04-21 at 10 39 47@2x

**Pull request merged**